### PR TITLE
New 1-wire component

### PR DESCRIPTION
**Pull request by ssieb**
# What does this implement/fix?

Moves the 1-wire support out of the dallas temperature sensor into its own platform.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3911

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [X] ESP8266
- [X] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
one_wire:
  - platform: gpio
    pin: 4

sensor:
  - platform: dallas_temp
    resolution: 12
    name: "Temperature"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by sphen13** [comment]
In my limited test here on a ESP32-WROOM-32D and a single DS18B20, this is working. It was not working previously and i had to use an external component.

**Comment by sphen13** [comment]
also tested same sensor on d1mini ESP8266 and it is working - some compilation warnings:

```INFO ESPHome 2024.5.4
INFO Reading configuration /config/esphome/dev-01.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing dev-01 (board: d1_mini; framework: arduino; platform: platformio/espressif8266@4.2.1)
--------------------------------------------------------------------------------
Library Manager: Installing esphome/noise-c @ 0.1.4
INFO Installing esphome/noise-c @ 0.1.4
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.4 has been installed!
INFO noise-c@0.1.4 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10018.1
INFO Installing esphome/libsodium @ 1.10018.1
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10018.1 has been installed!
INFO libsodium@1.10018.1 has been installed!
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
|-- noise-c @ 0.1.4
Compiling .pioenvs/dev-01/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/binary_sensor/automation.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/binary_sensor/binary_sensor.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/binary_sensor/filter.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/dallas_temp/dallas_temp.cpp.o
src/esphome/components/dallas_temp/dallas_temp.cpp: In member function 'virtual void esphome::one_wire::DallasTemperatureSensor::setup()':
src/esphome/components/dallas_temp/dallas_temp.cpp:89:29: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
   89 |   if (this->address_ & 0xff == DALLAS_MODEL_DS18S20) {
      |                        ~~~~~^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/dallas_temp/dallas_temp.cpp: In member function 'float esphome::one_wire::DallasTemperatureSensor::get_temp_c_()':
src/esphome/components/dallas_temp/dallas_temp.cpp:147:29: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
  147 |   if (this->address_ & 0xff == DALLAS_MODEL_DS18S20) {
      |                        ~~~~~^~~~~~~~~~~~~~~~~~~~~~~
Compiling .pioenvs/dev-01/src/esphome/components/duty_time/duty_time_sensor.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/esp8266/core.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/esp8266/gpio.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/esp8266/preferences.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/gpio/binary_sensor/gpio_binary_sensor.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/gpio/one_wire/gpio_one_wire.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/gpio/switch/gpio_switch.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/logger/logger_esp32.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/logger/logger_esp8266.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/logger/logger_host.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/logger/logger_libretiny.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/logger/logger_rp2040.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/mdns/mdns_esp32.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/mdns/mdns_host.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/mdns/mdns_libretiny.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/mdns/mdns_rp2040.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/one_wire/one_wire.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/one_wire/one_wire_bus.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/ota/ota_backend_arduino_esp32.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/ota/ota_backend_arduino_libretiny.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/ota/ota_backend_arduino_rp2040.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/ota/ota_component.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/script/script.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/sensor/automation.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/sensor/filter.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/sensor/sensor.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/sntp/sntp_component.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/socket/bsd_sockets_impl.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/socket/lwip_raw_tcp_impl.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/socket/lwip_sockets_impl.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/socket/socket.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/switch/automation.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/switch/switch.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/time/automation.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/time/real_time_clock.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/wifi/wifi_component_esp32_arduino.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/wifi/wifi_component_esp_idf.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/wifi/wifi_component_libretiny.cpp.o
Compiling .pioenvs/dev-01/src/esphome/components/wifi/wifi_component_pico_w.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/application.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/color.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/component.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/controller.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/log.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/ring_buffer.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/string_ref.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/time.cpp.o
Compiling .pioenvs/dev-01/src/esphome/core/util.cpp.o
Compiling .pioenvs/dev-01/src/main.cpp.o
Generating LD script .pioenvs/dev-01/ld/local.eagle.app.v6.common.ld
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiGratuitous.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/WiFiClient.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/WiFiClientSecureBearSSL.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/WiFiServer.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/WiFiServerSecureBearSSL.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/WiFiUdp.cpp.o
Compiling .pioenvs/dev-01/lib67b/ESP8266WiFi/enable_wifi_at_boot_time.cpp.o
Compiling .pioenvs/dev-01/lib3ff/ESP8266mDNS/ESP8266mDNS.cpp.o
Archiving .pioenvs/dev-01/lib67b/libESP8266WiFi.a
Compiling .pioenvs/dev-01/lib3ff/ESP8266mDNS/LEAmDNS.cpp.o
Compiling .pioenvs/dev-01/lib3ff/ESP8266mDNS/LEAmDNS_Control.cpp.o
Compiling .pioenvs/dev-01/lib3ff/ESP8266mDNS/LEAmDNS_Helpers.cpp.o
Compiling .pioenvs/dev-01/lib3ff/ESP8266mDNS/LEAmDNS_Structs.cpp.o
Compiling .pioenvs/dev-01/lib3ff/ESP8266mDNS/LEAmDNS_Transfer.cpp.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_aead/chacha20poly1305/sodium/aead_chacha20poly1305.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_core/ed25519/core_ed25519.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_core/ed25519/core_ristretto255.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_core/ed25519/ref10/ed25519_ref10.c.o
Archiving .pioenvs/dev-01/lib3ff/libESP8266mDNS.a
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_hash/crypto_hash.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_hash/sha256/cp/hash_sha256_cp.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_hash/sha256/hash_sha256.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_onetimeauth/crypto_onetimeauth.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_onetimeauth/poly1305/donna/poly1305_donna.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_onetimeauth/poly1305/onetimeauth_poly1305.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_scalarmult/crypto_scalarmult.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_scalarmult/curve25519/ref10/x25519_ref10.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_scalarmult/curve25519/scalarmult_curve25519.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_stream/chacha20/ref/chacha20_ref.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_stream/chacha20/stream_chacha20.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_stream/crypto_stream.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/crypto_verify/sodium/verify.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/sodium/codecs.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/sodium/core.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/sodium/runtime.c.o
Compiling .pioenvs/dev-01/lib54e/libsodium/sodium/utils.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/openssl/cipher-aesgcm.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/ref/cipher-aesgcm.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/ref/cipher-chachapoly.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/ref/dh-curve25519.c.o
Archiving .pioenvs/dev-01/lib54e/libsodium.a
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/ref/hash-blake2b.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/ref/hash-blake2s.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/ref/hash-sha256.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/sodium/cipher-aesgcm.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/sodium/cipher-chachapoly.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/sodium/dh-curve25519.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/sodium/hash-blake2b.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/backend/sodium/hash-sha256.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/aes/rijndael-alg-fst.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/blake2/blake2b.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/blake2/blake2s.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/chacha/chacha.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/donna/curve25519-donna-c64.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/donna/curve25519-donna.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/donna/poly1305-donna.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/sha2/sha256.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/sha2/sha512.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/crypto/x25519/x25519.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/cipherstate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/dhstate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/errors.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/handshakestate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/hashstate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/internal.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/names.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/patterns.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/rand_os.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/rand_sodium.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/randstate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/signstate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/symmetricstate.c.o
Compiling .pioenvs/dev-01/lib78b/noise-c/protocol/util.c.o
Archiving .pioenvs/dev-01/libFrameworkArduinoVariant.a
Compiling .pioenvs/dev-01/FrameworkArduino/Crypto.cpp.o
Archiving .pioenvs/dev-01/lib78b/libnoise-c.a
Compiling .pioenvs/dev-01/FrameworkArduino/Esp-frag.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Esp-version.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Esp.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/FS.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/FSnoop.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/FunctionalInterrupt.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/HardwareSerial.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/IPAddress.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/LwipDhcpServer-NonOS.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/LwipDhcpServer.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/LwipIntf.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/LwipIntfCB.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/MD5Builder.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Print.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Schedule.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/StackThunk.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Stream.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/StreamSend.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Tone.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/TypeConversion.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/Updater.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/WMath.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/WString.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/abi.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/aes_unwrap.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/base64.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/cbuf.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/cont.S.o
Compiling .pioenvs/dev-01/FrameworkArduino/cont_util.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_eboot_command.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_features.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_flash_quirks.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_flash_utils.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_i2s.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_main.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_non32xfer.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_noniso.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_phy.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_postmortem.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_si2c.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_sigma_delta.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_spi_utils.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_timer.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_vm.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_waveform_phase.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_waveform_pwm.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_wiring.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_wiring_analog.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_wiring_digital.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_wiring_pulse.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_wiring_pwm.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/core_esp8266_wiring_shift.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/crc32.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/debug.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/exc-c-wrapper-handler.S.o
Compiling .pioenvs/dev-01/FrameworkArduino/exc-sethandler.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/flash_hal.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/gdb_hooks.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/heap.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/hwdt_app_entry.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/libb64/cdecode.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/libb64/cencode.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/libc_replacements.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/mmu_iram.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/reboot_uart_dwnld.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/spiffs/spiffs_cache.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/spiffs/spiffs_check.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/spiffs/spiffs_gc.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/spiffs/spiffs_hydrogen.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/spiffs/spiffs_nucleus.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/spiffs_api.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/sqrt32.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/stdlib_noniso.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/time.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/uart.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/umm_malloc/umm_info.c.o
Compiling .pioenvs/dev-01/FrameworkArduino/umm_malloc/umm_integrity.c.o
Compiling .pioenvs/dev-01/FrameworkArduino/umm_malloc/umm_local.c.o
Compiling .pioenvs/dev-01/FrameworkArduino/umm_malloc/umm_malloc.cpp.o
Compiling .pioenvs/dev-01/FrameworkArduino/umm_malloc/umm_poison.c.o
Compiling .pioenvs/dev-01/FrameworkArduino/wpa2_eap_patch.cpp.o
Archiving .pioenvs/dev-01/libFrameworkArduino.a
Linking .pioenvs/dev-01/firmware.elf
RAM:   [====      ]  41.6% (used 34068 bytes from 81920 bytes)
Flash: [=====     ]  46.2% (used 482859 bytes from 1044464 bytes)
Building .pioenvs/dev-01/firmware.bin
esp8266_copy_factory_bin([".pioenvs/dev-01/firmware.bin"], [".pioenvs/dev-01/firmware.elf"])
========================= [SUCCESS] Took 35.22 seconds =========================
INFO Successfully compiled program.
```

**Comment by lboue** [comment]
Hello @ssieb. Thank you for your contribution.
I'm trying to test it and I'm getting an error:

```
INFO ESPHome 2024.6.0-dev
INFO Reading configuration /config/esphome/1-wire-bus.yaml...
Failed config

one_wire.gpio: [source /config/esphome/1-wire-bus.yaml:47]
  
  Platform not found: 'one_wire.gpio'.
  platform: gpio
  pin: 4
```

**ESPHome config**
```yaml
external_components:
    # https://github.com/esphome/esphome/pull/6860
  - source: github://pr#6860
    components: 
      - dallas
      - dallas_temp
      - gpio/one_wire
      - one_wire

# 1-Wire Bus
one_wire:
  - platform: gpio
    pin: 4

sensor:
  - platform: dallas_temp
    resolution: 12
    name: "Temperature"
```
Do you know how to correct this?

**Comment by ssieb** [comment]
Don't include `dallas` and it's just `gpio`.

**Comment by lboue** [comment]
> Don't include `dallas` and it's just `gpio`.

```yaml
external_components:
    # https://github.com/esphome/esphome/pull/6860
  - source: github://pr#6860
    components: 
      #- dallas
      #- dallas_temp
      - gpio/one_wire
      - one_wire

# The gpio platform uses the CPU to generate the bus signals on a GPIO pin.
one_wire:
  - platform: gpio
    pin: 4
```
This generates the same error:
```
Download 1-wire-bus.yaml
INFO ESPHome 2024.6.0-dev
INFO Reading configuration /config/esphome/1-wire-bus.yaml...
Failed config

one_wire.gpio: [source /config/esphome/1-wire-bus.yaml:49]
  
  Platform not found: 'one_wire.gpio'.
  platform: gpio
  pin: 4
```

**Comment by ssieb** [comment]
You do need the `dallas_temp` and `gpio`, not `gpio/one_wire`.

**Comment by lboue** [comment]
Thanks. It's OK this time.

```yaml
external_components:
    # https://github.com/esphome/esphome/pull/6860
  - source: github://pr#6860
    components: 
      - dallas_temp
      - gpio
      - one_wire
    refresh: 0s

one_wire:
  - platform: gpio
    pin: 4

# DS18B20 Temperature Sensor
sensor:
  - platform: dallas_temp
    resolution: 12
    name: "Temperature"
```

File: tests/components/dallas/common.yaml
Comment (jesserockz) [inline tests/components/dallas/common.yaml:1]: This folder should be renamed to `dallas_temp` or it will not be run

File: tests/test1.yaml
Comment (jesserockz) [inline tests/test1.yaml:1]: Lets just drop the dallas/one-wire from the old test files

File: esphome/components/dallas_temp/sensor.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+
+async def to_code(config):
+    var = await sensor.new_sensor(config)
+    await cg.register_component(var, config)
+    await one_wire.register_one_wire_device(var, config)
+
+    if CONF_RESOLUTION in config:
+        cg.add(var.set_resolution(config[CONF_RESOLUTION]))
```
Comment (jesserockz) [inline esphome/components/dallas_temp/sensor.py]: There is a default so checking is not needed
```suggestion
    cg.add(var.set_resolution(config[CONF_RESOLUTION]))
```

File: esphome/components/one_wire/__init__.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    """Register an 1-wire device with the given config.
+
+    Sets the 1-wire bus to use and the 1-wire address.
+
+    This is a coroutine, you need to await it with a 'yield' expression!
+    """
+    parent = await cg.get_variable(config[CONF_ONE_WIRE_ID])
+    cg.add(var.set_one_wire_bus(parent))
+    if CONF_ADDRESS in config:
+        cg.add(var.set_address(config[CONF_ADDRESS]))
```
Comment (jesserockz) [inline esphome/components/one_wire/__init__.py]:
```suggestion
    if (address := config.get(CONF_ADDRESS)) is not None:
        cg.add(var.set_address(address))
```

File: esphome/components/dallas_temp/dallas_temp.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    float tempc = this->get_temp_c_();
+    ESP_LOGD(TAG, "'%s': Got Temperature=%.1f°C", this->get_name().c_str(), tempc);
+    this->publish_state(tempc);
+  });
+}
+
+void IRAM_ATTR DallasTemperatureSensor::read_scratch_pad_int_() {
+  for (unsigned char &i : this->scratch_pad_) {
```
Comment (jesserockz) [inline esphome/components/dallas_temp/dallas_temp.cpp]:
```suggestion
  for (uint8_t &i : this->scratch_pad_) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Comment by ssieb** [comment]
Release note:
The `dallas` component has been split up into the `one_wire` bus and the `dallas_temp` sensor.  This will allow for other 1-wire interfaces to be added and other 1-wire devices that aren't temperature sensors.  The `index` option has been removed.  If there is only one device on the bus, then the address is optional.  Otherwise, you must specify the address.

**Pull request merged**